plugins/secom_cl.py

```diff
@@ -196,322 +221,463 @@ def sem_roi_to_ccd(escan, ccd, roi, margin=0):
+        scale = numpy.array([pxs[0] / spxs[0], pxs[1] / spxs[1]])
+        min_scale = numpy.array(self._scanner.scale.range[0])  # fixed for scanner ?
+        max_scale = numpy.array([self._scanner.shape[0], self._scanner.shape[1]])  # fixed for scanner ?
+        scale = numpy.maximum(min_scale, numpy.minimum(scale, max_scale))
+        pxs = tuple(scale * numpy.array(spxs))  # tuple (x, y)
+
+        # compute the repetition (ints) that fits the ROI with the pixel size
+        sshape = self._scanner.shape
+        roi_size = (roi[2] - roi[0], roi[3] - roi[1])
+        rep = (int(round(sshape[0] * roi_size[0] / scale[0])),
+               int(round(sshape[1] * roi_size[1] / scale[1])))
+
+        logging.debug("First trial with roi = %s, rep = %s, pxs = %s", roi, rep, pxs)
+
+        # Ensure it's really compatible with the hardware
+        rep = self._scanner.resolution.clip(rep)
+
+        # update the ROI so that it's _exactly_ pixel size * repetition,
+        # while keeping its center fixed
+        roi = self._adaptROI2(roi, rep, pxs)
+        roi = self._fitROI(roi)
+
+        # In case the ROI got modified again and the aspect ratio is not anymore
+        # the same as the rep, we shrink it to ensure the pixels are square (and
+        # it should still fit within the FoV).
+        eratio = sshape[0] / sshape[1]
+
+        # Double check we didn't end up with scale out of range
+        pxs_range = self._getPixelSizeRange2()
+        if not pxs_range[0] <= pxs[0] <= pxs_range[1] or not pxs_range[0] <= pxs[1] <= pxs_range[1]:
+            logging.error("Computed impossibly small pixel size %s, with range %s", pxs, pxs_range)
+            # TODO: revert to some *acceptable* values for ROI + rep + PxS?
+            # pxs = max(pxs_range[0], min(pxs, pxs_range[1]))
+
+        logging.debug("Computed roi = %s, rep = %s, pxs = %s", roi, rep, pxs)
+
+        return tuple(roi), tuple(rep), tuple(pxs)
+
+    def _adaptROI2(self, roi, rep, pxs):
+        """
+        Compute the ROI so that it's _exactly_ pixel size * repetition,
+        while keeping its center fixed
+        :param roi: (4 floats) The current ROI, just to know its center.
+        :param rep: (2 ints)
+        :param pxs: (float)
+        :returns: ROI (4 floats) ltrb
+        """
+        # Rep + PxS (+ center of ROI) -> ROI
+        roi_center = ((roi[0] + roi[2]) / 2,
+                      (roi[1] + roi[3]) / 2)
+        spxs = self._scanner.pixelSize.value
+        sshape = self._scanner.shape
+        phy_size = (spxs[0] * sshape[0], spxs[1] * sshape[1])  # max physical ROI
+        roi_size = (rep[0] * pxs[0] / phy_size[0],
+                    rep[1] * pxs[1] / phy_size[1])
+        roi = (roi_center[0] - roi_size[0] / 2,
+               roi_center[1] - roi_size[1] / 2,
+               roi_center[0] + roi_size[0] / 2,
+               roi_center[1] + roi_size[1] / 2)
+
+        return roi
+
+    def _fitROI2(self, roi):
```

Comment:
I don't think you need to duplicate that one, do you?

Reply:
Yes I think we need it, because `roi, rep, pxs = self._updateROIAndPixelSize2(roi, pxs)` is called somewhere in this method.